<!--REVIEW_SCORE-->
MAJOR=3
MINOR=1
SCORE=32
VERDICT=BLOCKED
TESTS=FAIL
<!--/REVIEW_SCORE-->

## Code Review — PR #67

**Reviewer**: REVIEWER agent (automated, GPT 5.3 Codex)
**Verdict**: BLOCKED
**Score**: 32 (MAJOR=3 × 10 + MINOR=1 × 2)

### MAJOR Findings (blocks merge)
1. **TEST-03** — `test-run`: `npm test` fails globally (`Cannot find module 'expo-modules-core' from jest-expo preset`), so the PR cannot be merge-gated as passing tests. **Fix**: restore a working Jest/Expo test runtime so `npm test` passes.
2. **DATA-INT-01** — `lib/db/achievements.ts:1382`: `saveEarnedAchievements()` performs multi-row writes in a loop without a transaction. This can partially persist achievements on interruption. **Fix**: wrap the insert loop in `db.withTransactionAsync()` (or existing transaction helper) for atomicity.
3. **ERR-DB-01** — `app/progress/achievements.tsx:489` and `app/session/summary/[id].tsx:818`: DB/evaluation errors are swallowed (`catch {}`) without user-facing error messaging. **Fix**: surface a visible error state/toast consistent with app patterns when achievement loading/evaluation fails.

### MINOR Findings (advisory)
1. **TEST-02** — Test coverage is incomplete for the primary changed feature surface: `__tests__/achievements.test.ts` covers pure evaluator logic, but there is no coverage for DB integration (`lib/db/achievements.ts`) or new UI flows (`app/progress/achievements.tsx`, summary achievement card).

### Deterministic Check Results
CRITICAL: 0 FAIL, 7 PASS  
MAJOR: 0 FAIL, 2 PASS  
Total: 0 FAIL, 9 PASS

### Test Results
FAIL: 56 failed, 56 total (Jest runtime error: missing `expo-modules-core` in `jest-expo` preset)

--- a/app/_layout.tsx
+++ b/app/_layout.tsx
@@ -315,6 +315,12 @@ export default function RootLayout() {
               }}
             />
             <Stack.Screen
+              name="progress"
+              options={{
+                headerShown: false,
+              }}
+            />
+            <Stack.Screen
               name="history"
               options={{
                 headerShown: true,

--- a/app/progress/achievements.tsx
+++ b/app/progress/achievements.tsx
@@ -50,6 +50,7 @@ export default function AchievementsScreen() {
   const [items, setItems] = useState<AchievementItem[]>([]);
   const [earnedCount, setEarnedCount] = useState(0);
   const [loading, setLoading] = useState(true);
+  const [error, setError] = useState<string | null>(null);
   const [retroBanner, setRetroBanner] = useState<number | null>(null);
 
   useFocusEffect(
@@ -80,6 +81,7 @@ export default function AchievementsScreen() {
 
           // Show retroactive banner on first open
           const seenBanner = await hasSeenRetroactiveBanner();
+          if (cancelled) return;
           if (!seenBanner && earnedMap.size > 0) {
             setRetroBanner(earnedMap.size);
             await markRetroactiveBannerSeen();
@@ -101,8 +103,9 @@ export default function AchievementsScreen() {
             })),
           );
           setEarnedCount(earnedMap.size);
-        } catch {
-          // Evaluation failed — show empty state with error
+        } catch (e) {
+          console.warn("Achievement evaluation failed:", e);
+          if (!cancelled) setError("Could not load achievements. Please try again later.");
         } finally {
           if (!cancelled) setLoading(false);
         }
@@ -129,6 +132,20 @@ export default function AchievementsScreen() {
         <Stack.Screen options={{ title: "Achievements" }} />
         <View style={[styles.center, { backgroundColor: theme.colors.background }]}>
           <Text style={{ color: theme.colors.onSurfaceVariant }}>Loading achievements...</Text>
+        </View>
+      </>
+    );
+  }
+
+  if (error) {
+    return (
+      <>
+        <Stack.Screen options={{ title: "Achievements" }} />
+        <View style={[styles.center, { backgroundColor: theme.colors.background }]}>
+          <Text variant="headlineMedium" style={{ marginBottom: 8 }}>⚠️</Text>
+          <Text variant="bodyLarge" style={{ color: theme.colors.onSurfaceVariant, textAlign: "center", padding: 16 }}>
+            {error}
+          </Text>
         </View>
       </>
     );
@@ -172,7 +189,7 @@ export default function AchievementsScreen() {
             ? `${item.name} achievement — Earned on ${formatDate(item.earnedAt!)}`
             : `${item.name} achievement — Locked, ${Math.round(item.progress * 100)}% complete`
         }
-        accessibilityRole="button"
+        accessibilityRole="summary"
       >
         <Card.Content style={styles.badgeContent}>
           <View style={styles.iconContainer}>

--- a/app/session/summary/[id].tsx
+++ b/app/session/summary/[id].tsx
@@ -97,8 +97,8 @@ export default function Summary() {
             Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Medium).catch(() => {});
           }
         }
-      } catch {
-        // Achievement evaluation is non-critical
+      } catch (e) {
+        console.warn("Achievement evaluation failed:", e);
       }
 
       AccessibilityInfo.announceForAccessibility("Workout Complete!");

--- a/lib/db/achievements.ts
+++ b/lib/db/achievements.ts
@@ -1,7 +1,7 @@
 // Database queries for the achievement system.
 // Builds AchievementContext from batched queries and manages achievements_earned table.
 
-import { query, queryOne, execute } from "./helpers";
+import { query, queryOne, execute, withTransaction } from "./helpers";
 import type { AchievementContext, EarnedAchievement } from "../achievements";
 
 // --- Context Building ---
@@ -129,12 +129,14 @@ export async function saveEarnedAchievements(
   achievementIds: string[],
   earnedAt: number = Date.now(),
 ): Promise<void> {
-  for (const id of achievementIds) {
-    await execute(
-      "INSERT OR IGNORE INTO achievements_earned (achievement_id, earned_at) VALUES (?, ?)",
-      [id, earnedAt]
-    );
-  }
+  await withTransaction(async (db) => {
+    for (const id of achievementIds) {
+      await db.runAsync(
+        "INSERT OR IGNORE INTO achievements_earned (achievement_id, earned_at) VALUES (?, ?)",
+        [id, earnedAt]
+      );
+    }
+  });
 }
 
 export async function getEarnedCount(): Promise<number> {

